src/sparse/linalg/ordering.rs

```diff
@@ -13,74 +11,512 @@ pub struct Ordering<I> {
+///
+/// When calling this function, all generic parameters need to specified.
+pub fn cuthill_mckee_custom<N, I, Iptr, S, D>(
+    mat: CsMatViewI<N, I, Iptr>,
+    mut starting_strategy: S,
+    mut directed_ordering: D,
+) -> Ordering<I>
 where
     N: PartialEq,
     I: SpIndex,
     Iptr: SpIndex,
+    S: start::Strategy<N, I, Iptr>,
+    D: order::DirectedOrdering<I>,
 {
     debug_assert!(is_symmetric(&mat));
     assert_eq!(mat.cols(), mat.rows());
+
     let nb_vertices = mat.cols();
+    let degrees = mat.degrees();
+    let max_neighbors = degrees.iter().max().cloned().unwrap_or(0);
+
+    // This will be transformed into the actual `Ordering` in the end,
+    // contains the permuntation and component delimeters.
+    directed_ordering.prepare(nb_vertices);
+
+    // This is the 'working data', into which new neighboring, sorted vertices are inserted,
+    // the next vertex to process is popped from here.
     let mut deque = VecDeque::with_capacity(nb_vertices);
-    let max_neighbors = mat
-        .indptr()
-        .windows(2)
-        .map(|w| w[1] - w[0])
-        .max()
-        .unwrap_or(Iptr::zero());
-    let mut neighbors = Vec::with_capacity(max_neighbors.index());
-    let mut perm = Vec::with_capacity(nb_vertices);
+
+    // This are all new neighbors of the currently processed vertex, they are collected here
+    // to be sorted prior to being appended to 'deque'.
+    // The alternative of immediately pushing to deque and sorting there surprisingly performs worse.
+    let mut neighbors = Vec::with_capacity(max_neighbors);
+
+    // Storing which vertices have already been visited.
     let mut visited = vec![false; nb_vertices];
-    let mut connected_parts = Vec::with_capacity(4);
-    connected_parts.push(0);
 
-    let degrees = mat.degrees();
+    for perm_index in 0..nb_vertices {
+        // Find the next index to process, choosing a new starting vertex if necessary.
+        let current_vertex = deque.pop_front().unwrap_or_else(|| {
+
+            // We found a new connected component, starting at this iteration.
+            directed_ordering.add_component_delimeter(perm_index);
 
-    while perm.len() < nb_vertices {
-        // find the non-visited vertex with the lowest degree
-        let mut min_deg = nb_vertices;
-        let mut min_deg_vert = 0;
-        for (vert, vis) in visited.iter().enumerate() {
-            let vert_deg = degrees[vert];
-            if !vis && vert_deg <= min_deg {
-                min_deg = vert_deg;
-                min_deg_vert = vert;
+            // Find a new starting vertex, using the given strategy.
+            let new_start_vertex = starting_strategy.find_start_vertex(&visited, &degrees, &mat);
+            assert!(!visited[new_start_vertex], "Vertex returned by starting strategy should always be unvisited");
+
+            new_start_vertex
+        });
+
+        // Write the next permutation in reverse order.
```

Comment:
This comment needs to be updated I guess now that the order is custom.

Reply:
Yes.